### ISSUE TITLE
(WIP) FR: make local recursive workers an optimisation again with "push down lets" (static argument transformation)

### ISSUE THREAD

**Issue by brprice** [hackworthltd/primer#1112]
Thank you for helping us improve our software!

**Before filing your feature request, please read and acknowledge the following by checking all the boxes:**

- [X] I have read the project's [code of conduct](https://github.com/hackworthltd/primer/blob/main/CODE_OF_CONDUCT.md), and by submitting this feature request, I agree to follow it.

## Description

This is a (work-in-progress) feature request. I am splitting it out of #736 as a possible future enhancement

A side effect of #736 was that writing local recursive helpers that "bake out" arguments that do not change in each recursive call is now actually worse. E.g. consider the two definitions
```
  map,mapSAT :: (a -> b) -> [a] -> [b]
  map f xs = case xs of
               [] -> []
               y : ys -> f y : map f ys
  mapSAT f = let go xs = case xs of
                           [] -> [] 
                           y : ys -> f y : go ys
             in go
```
Previously, `mapSAT` would take fewer steps (in the normal-order EvalFull, but also quite likely in a student-led interactive eval session), and have smaller trees. However, after pushing down lets, this is no longer the case.

## Detailed explanation of the issue
For now, I quote a note in the source:
```
{- Note [Pushing down lets and the static argument transformation]
Our strategy of "pushing down lets" (which is effectively an explicit
substitution semantics, where the substitution is interspersed with
other reduction steps) has different performance characteristics than
our previous strategy of eagerly fully inlining `let` bindings. In
particular, it makes the transformation of unit_8 into unit_9 is not
effective.

To explain why, let us first recap what this transformation is.
The static argument transformation is a technique of optimising
a recursive function of multiple arguments by splitting the lambda
bindings into two: the ones that do not change in a recursive call, and
those that do. Then the changing ones are handled by a local (recursive)
function. For example, changing the following definition of `map`
  map :: (a -> b) -> [a] -> [b]
  map f xs = case xs of
               [] -> []
               y : ys -> f y : map f ys
into
  mapSAT f = let go xs = case xs of
                           [] -> [] 
                           y : ys -> f y : go ys
             in go

When reducing by naively applying rewrite rules and doing "long-range"
substitution, the latter can be more efficient, since we effectively
specialise the recursive loop for a particular `f`, rather than passing
it around each iteration. (Note that passing it around causes an extra
beta step on each recursive call, and thus more substitution.) That is
to say, calling `mapSAT foo [1,2,3]` will result in
  let go xs = case xs of
                [] -> []
                y : ys -> foo y : go ys
  in go [1,2,3]

However, that is assuming that beta reductions (and `let`s, since we
reduce betas to let bindings) result in "eager substitution":
  let f = foo in
    let go = c (...) (...) (... f ... go ...) in
      go
should reduce by inlining the `f` inside the recursive let binding.

When we lazily push an explicit substitution down the tree this does not
happen (at least, in our implementation, since we treat the recusive let
as a substition and never push substitutions into each other). Indeed,
we would reduce the above by inlining the `go`
  let f = foo in
    let go = c (...) (...) (... f ... go ...) in
      ... f ... go ...
and then pushing the `let`s (i.e. explicit substitutions) down, resulting in
  c (...)
    (...)
    (let f = foo in
       let go = c (...) (...) (... f ... go ...) in
         ... f ... go ...)
(Here we have taken the optimization of not pushing the substition into
branches where those variables are unused.) Notice the recursive structure
here will hold on to the substitution, pushing it down into each expansion
of the recursive let, which is no better than the original definition
without the static argument translation's local worker.

It may be possible to avoid this by treating recursive bindings
differently to explicit substitutions, so we could push the substitution
of `f` into the definition of `go`. However, I have not thought about
this enough (from any of a theoretical, implementation or pedagogy
standpoint) to have a good idea of whether it would work and be worth
it. In particular, for a simple language aimed at learning, do we
want to treat non-recursive and recursive lets so differently (or
maybe even user-written lets vs explicit substitutions caused by beta
reduction)? Alternatively, it may also be possible to cause explicit
substitutions to interact with let bindings (i.e. push the outer let
rather than the inner let, and not special-case any particular lets),
but a naive version of this would cause infinite loops continually
swapping two lets!
-}
```
and also my explanation in the #736 FR
```
        -- note that ww/map is somewhat pointless when push down lets -- maybe we should push lets inside letrecs (maybe just inside type?)?
        --     nb, get something like @let α=Char, β=Char, f=toUpper in (letrec go : List α -> List β; go = λxs.RHS in go)
        --     and then in two steps (expand @go@, push stack of let+letrec)
        --        @λxs. let α=Char, β=Char, f=toUpper in (letrec go : List α -> List β; go = λxs.RHS in RHS : List α -> List β)
        --     we carry around the subst for α,β and f, using α,β inside annotation and f in RHS each time expand the letrec
```
To elaborate on the above:

Recall that the w/w map is
```
map :: ∀a. ∀b. (a -> b) -> List a -> List b
map = Λa. Λb. λf.
    letrec go :: List a -> List b
           go = λxs. case xs of
                       Nil -> Nil
                       Cons y ys -> Cons (f y) (go ys)
    in go
```
Now consider the application `map Char Char toUpper "abc"`, we get via inlining the global and one BETA step
```
((let a = Char in Λb. λf.
    letrec go :: ...
           go = ...
    in go)
 :: let a = Char in ∀b.(a -> b) -> List a -> List b) Char toUpper "abc"
```
and then we fully reduce the type annotation, before looking at the term (yes this is a bit odd, see my other comment)
```
((let a = Char in Λb. λf.
    letrec go :: ...
           go = ...
    in go)
 :: ∀b.(Char -> b) -> List Char -> List b) Char toUpper "abc"
```
Now one PushLet gives
```
((Λb. let a = Char in λf.
    letrec go :: ...
           go = ...
    in go)
 :: ∀b.(Char -> b) -> List Char -> List b) Char toUpper "abc"
```
which is very similar to the above. Another BETA and reduction of the annotation gives
```
((let b = Char in let a = Char in λf.
    letrec go :: ...
           go = ...
    in go)
 :: (Char -> Char) -> List Char -> List Char) toUpper "abc"
```
Now push the lets again
```
((λf. let b = Char in let a = Char in
    letrec go :: ...
           go = ...
    in go)
 :: (Char -> Char) -> List Char -> List Char) toUpper "abc"
```
And a Beta step
```
((let f = (toUpper :: Char -> Char) in let b = Char in let a = Char in
    letrec go :: ...
           go = ...
    in go)
 :: List Char -> List Char) "abc"
```
Now (and this is the first step of what I above mentioned as "in two steps") we expand the `go`
```
((let f = (toUpper :: Char -> Char) in let b = Char in let a = Char in
    letrec go :: ...
           go = λxs. ...
    in λxs. ...)
 :: List Char -> List Char) "abc"
```
and push the lets
```
((λxs. let f = (toUpper :: Char -> Char) in let b = Char in let a = Char in
    letrec go :: List a -> List b
           go = λxs. ... Cons (f y) (go ys)
    in case ...)
 :: List Char -> List Char) "abc"
```
Note that this will now reduce by Beta-reducing the `λxs`, and we will keep the stack of lets `f,b,a` around because they are used in the annotation and body of the letrec.
Before we changed to pushing-down-lets, we would have inlined `f,b,a`, giving `letrec go :: List Char -> List Char; go = λxs. case xs of Nil -> Nil ; Cons y ys -> Cons (toUpper y) (go ys)`, which is smaller and faster.

<details>
<summary>Some selected steps in the rest of the reduction, eliding some annotations:</summary>

```
let xs = "abc" in let f = (toUpper :: Char -> Char) in let b = Char in let a = Char in
    letrec go :: List a -> List b
           go = λxs. ... Cons (f y) (go ys)
    in case ...
```
```
case (let xs = "abc" in xs) of
  Nil -> Nil
  Cons y ys -> let f = (toUpper :: Char -> Char) in let b = Char in let a = Char in
               letrec go :: List a -> List b
                      go = λxs. case xs of
                                  Nil -> Nil
                                  Cons y ys -> Cons (f y) (go ys)
               in Cons (f y) (go ys)
```
```
case "abc" of
  Nil -> Nil
  Cons y ys -> let f = (toUpper :: Char -> Char) in let b = Char in let a = Char in
               letrec go :: List a -> List b
                      go = λxs. case xs of
                                  Nil -> Nil
                                  Cons y ys -> Cons (f y) (go ys)
               in Cons (f y) (go ys)
```
```
let y='a' in let ys="bc" in
let f = (toUpper :: Char -> Char) in let b = Char in let a = Char in
letrec go :: List a -> List b
       go = λxs. case xs of
                    Nil -> Nil
                    Cons y ys -> Cons (f y) (go ys)
in Cons (f y) (go ys)
```
```
Cons (let y='a' in let f = toUpper in f y)
     (let ys="bc" in
      let f = (toUpper :: Char -> Char) in let b = Char in let a = Char in
      letrec go :: List a -> List b
             go = λxs. case xs of
                         Nil -> Nil
                         Cons y ys -> Cons (f y) (go ys)
      in go ys)
```
```
Cons 'A'
     ((let f = (toUpper :: Char -> Char) in let b = Char in let a = Char in
       letrec go :: List a -> List b
              go = λxs. case xs of
                          Nil -> Nil
                          Cons y ys -> Cons (f y) (go ys)
       in go)
      (let ys="bc" in ys))
```
```
Cons 'A'
  (Cons 'B'
     ((let f = (toUpper :: Char -> Char) in let b = Char in let a = Char in
       letrec go :: List a -> List b
              go = λxs. case xs of
                          Nil -> Nil
                          Cons y ys -> Cons (f y) (go ys)
       in go)
      (let ys="c" in ys)))
```
```
Cons 'A'
  (Cons 'B'
    (Cons 'C'
     ((let f = (toUpper :: Char -> Char) in let b = Char in let a = Char in
       letrec go :: List a -> List b
              go = λxs. case xs of
                          Nil -> Nil
                          Cons y ys -> Cons (f y) (go ys)
       in go)
      (let ys=Nil in ys))))
```
```
Cons 'A'
  (Cons 'B'
    (Cons 'C'
      (case (let xs = Nil in xs) of
         Nil -> Nil
         Cons y ys -> let f = (toUpper :: Char -> Char) in let b = Char in let a = Char in
                      letrec go :: List a -> List b
                             go = λxs. case xs of
                                         Nil -> Nil
                                         Cons y ys -> Cons (f y) (go ys)
                      in Cons (f y) (go ys))))
```
```
Cons 'A' (Cons 'B' (Cons 'C' Nil))
```
</details>

## Dependencies

#736 
<*Link to other GitHub Issues or PRs that must be completed before work can begin on this feature.*>

## Spec

There is no actual plan as yet. If we decide to change anything here, we plan to elaborate this into a full feature request.
One half-baked idea is simply to allow explicit substitutions to push inside recursive lets: `let x=e in (letrec f = s[x,f] in t[x,f])  ~>  letrec f = (let x=e in s[x,f]) in (let x = e in t[x,f])`, or similar. However, note that I have not yet convinced myself this is a good idea (or correct!).

<*Describe what the feature will do, from a high-level perspective.*>

## Implementation details

<*If there are any particularly tricky implementation bits that are worth discussing or you haven't quite figured out yet, describe the details here.*>

## Not in spec

<*Describe any features that your implementation is explicitly avoiding, that a reasonable person might think should be in spec. For example, if you're adding a new action that operates on variables, but it only works with term variables and not type variables, you might want to mention it here so that the scope of the feature is clear.*>

## Discussion

<*If there's a GitHub Discussions topic or a GitHub Issue where this feature has been/is being discussed, link it here.*>

## Future work

<*Describe ideas or additional features that might be useful once this feature has been implemented. This is a good place to link to other GitHub Issues or PRs that track this future work.*>
